### PR TITLE
engine: when parsing/unparsing dockerfile to inject image deps, preserve '# syntax =' flag

### PR DESCRIPTION
**Pull request by maiamcc**
Hello @nicks, @landism,

Please review the following commits I made in branch maiamcc/dockerfile-syntax-flag:

34ae0c55498fb2a380bfb5568a340e24d8303bca (2019-10-18 13:49:28 -0400)
engine: when parsing/unparsing dockerfile to inject image deps, preserve '# syntax =' flag

Addresses #2371. Verified on [this example branch](https://github.com/windmilleng/servantes/tree/experimental-dockerfile-child-image).

I'm gonna file an issue with buildkit about pulling in syntax flags during parsing, tho idk if they'll care.

### CODE REVIEW

File: internal/dockerfile/dockerfile.go [hunk 1]
```diff
@@ -62,6 +69,17 @@ func (d Dockerfile) RmPaths(pathsToRm []string) Dockerfile {
 	return d.Join(fmt.Sprintf("RUN %s", rmCmd.String()))
 }
 
+// Pull out the `# syntax = docker/xxx` line, if one exists.
+func (d Dockerfile) MaybeSyntaxFlag() Dockerfile {
```
Comment (nicks) [inline internal/dockerfile/dockerfile.go]: this is not quite right. can you use the docker parsing code for this?
https://github.com/moby/buildkit/blob/3f8ab160d5079539f9ed971fb069e4205108dd9d/frontend/dockerfile/dockerfile2llb/directives.go#L27

File: internal/engine/image_build_and_deployer.go [hunk 2]
```diff
@@ -391,6 +392,13 @@ func injectImageDependencies(iTarget model.ImageTarget, iTargetMap map[model.Tar
 		return model.ImageTarget{}, errors.Wrap(err, "injectImageDependencies")
 	}
 
+	// HACK(maia): until buildkit/parse supports `# syntax = ` flags, check whether the original
```
Comment (nicks) [inline internal/engine/image_build_and_deployer.go]: this feels like the wrong place to put this code. Can you move this into the ast and Print() function?

Comment (maiamcc) [reply internal/engine/image_build_and_deployer.go]: fair, will do

File: internal/dockerfile/dockerfile.go (hunk 1 above)
Comment (maiamcc) [reply internal/dockerfile/dockerfile.go]: amazing, i didn't realize that existed!

File: internal/dockerfile/ast.go [hunk 3]
```diff
@@ -113,6 +116,17 @@ func (a AST) traverseNode(node *parser.Node, visit func(*parser.Node) error) err
 func (a AST) Print() (Dockerfile, error) {
 	buf := bytes.NewBuffer(nil)
 	currentLine := 1
+
+	directiveFmt := "# %s = %s\n"
+	for k, v := range a.directives {
+		dir := fmt.Sprintf(directiveFmt, k, v)
+		_, err := buf.Write([]byte(dir))
```
Comment (nicks) [inline internal/dockerfile/ast.go]: you can collapse these two lines with an Fprintf

**Pull request merged**